### PR TITLE
Contingency

### PR DESCRIPTION
**Pull request by rbolgaryn**
implements contingency evaluation functions (N-1 analysis)

### CODE REVIEW

**Review by SteffenMeinecke — COMMENTED** [review]
seems that you are still working on the tutorial, so:

* fix the ending of the 1st sentence in the last section

Three questions, which I guess are considered, but just to be sure:

1. With `res = pp.contingency.run_contingency(net, nminus1_cases)`, `res["line"]["loading_percent"]` represents the n-0 results (similar for bus results). Is it somewhere documented? (I didn't find it in the docstring.)
2. It should be mentioned what kind of results would be set to the `res_` tables if `write_to_net is True`
3. Is `net.line.max_loading_percent` considered for `res["line"]["cause_overloadings"]`? I guess yes but wanted to ask for sure...

**Comment by rbolgaryn** [comment]
tests do not pass fpr Python 3.7 - maybe due to the numpy version

**Review by SteffenMeinecke — DISMISSED** [review]
Could you answer the 3 questions above and integrate documentation improvements if resulting from that?

**Comment by rbolgaryn** [comment]
> seems that you are still working on the tutorial, so:
> 
>     * fix the ending of the 1st sentence in the last section
> 
> 
> Three questions, which I guess are considered, but just to be sure:
> 
>     1. With `res = pp.contingency.run_contingency(net, nminus1_cases)`, `res["line"]["loading_percent"]` represents the n-0 results (similar for bus results). Is it somewhere documented? (I didn't find it in the docstring.)
> 
>     2. It should be mentioned what kind of results come within the `res_` tables after contingency analysis with `write_to_net is True`. Are these the n-0 results?
> 
>     3. Is `net.line.max_loading_percent` considered when evaluating `res["line"]["cause_overloadings"]`? I guess yes but wanted to ask for sure...

Hi Steffen, here the answers:

- tutorial is updated
- the "loading_percent" variable in the results is for N-0 case (updated docstrings)
- added more information in the docstrings about what results are returned after contingency calculation
- net.line.max_loading_percent is considered. If provided, net.max_loading_percent_nminus1 is considered for N-1 (added in docstrings)

**Comment by SteffenMeinecke** [comment]
> Hi Steffen, here the answers:

Perfect, then please merge.

File: pandapower/contingency/contingency.py [hunk 1]
```diff
@@ -0,0 +1,489 @@
+# -*- coding: utf-8 -*-
+
+# Copyright (c) 2016-2022 by University of Kassel and Fraunhofer Institute for Energy Economics
```
Comment (SteffenMeinecke) [inline pandapower/contingency/contingency.py:3]: copyright 2016-2023 would be better

Comment (rbolgaryn) [reply pandapower/contingency/contingency.py:3]: oops

File: pandapower/test/contingency/test_contingency.py [hunk 2]
```diff
@@ -0,0 +1,390 @@
+# -*- coding: utf-8 -*-
+
+# Copyright (c) 2016-2022 by University of Kassel and Fraunhofer Institute for Energy Economics
```
Comment (SteffenMeinecke) [inline pandapower/test/contingency/test_contingency.py]: copyright year: same here

File: pandapower/test/contingency/test_contingency.py [hunk 3]
```diff
@@ -0,0 +1,390 @@
...
+    pp.contingency.run_contingency_ls2g(net, nminus1_cases,
+                                        contingency_evaluation_function=run_for_from_bus_loading)
+    res = {"trafo": net.res_trafo.copy(), "line": net.res_line.copy()}
+    for element, val in nminus1_cases.items():
+        for i in val["index"]:
+            net[element].at[i, "in_service"] = False
+            run_for_from_bus_loading(net)
+            net[element].at[i, 'in_service'] = True
+            idx_overloaded_tr = net.res_trafo.loc[net.res_trafo.loading_percent > net.trafo.max_loading_percent].index
+            congestion_tr = (net.res_trafo.loc[idx_overloaded_tr, 'loading_percent'].values - net.trafo.loc[idx_overloaded_tr, 'max_loading_percent'].values) * net.trafo.loc[idx_overloaded_tr, 'sn_mva'].values / 100
```
Comment (SteffenMeinecke) [inline pandapower/test/contingency/test_contingency.py:230]: this line is very long

**Review by SteffenMeinecke — DISMISSED** [review]

**Review by SteffenMeinecke — APPROVED** [review]

**Pull request merged**